lib/tree.rb

```diff
@@ -16,42 +16,141 @@ def initialize
+#otherwise if node is not empty, then set it as the key
+#if key is less than -> left
+#if key is greater than -> right
+def find_helper(node, key)
+  if node.nil?
+    return nil
+  elsif node.key == key
+    return node.value
+  elsif key < node.key
+    return find_helper(node.left, key)
+  elsif key > node.key
+    return find_helper(node.right, key)
+  end
+end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    return find_helper(@root, key)
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  #--------------------------------------------------------------------------#
+  #use list is a reference to pass
+  #check if node is empty, if yes, return the list
+  def inorder_helper(node, list) 
+    if node.nil?
+      return list
+    end
+    #push key,value
+    list.push({key: node.key, value: node.value})
+    inorder_helper(node.left, list)
+    inorder_helper(node.right, list)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  #--------------------------------------------------------------------------#
+  #used same logic as inorder
+  #use list is a reference to pass
+  #check if node is empty, if yes, return the list
+  def preorder_helper(node, list) 
+    if node.nil?
+      return list 
+    end
+
+    #push key,value 
+    list.push({key: node.key, value: node.value})
+    preorder_helper(node.left, list)
+    preorder_helper(node.right, list)
+  end
+  
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  #--------------------------------------------------------------------------#
+    #used same logic as inorder
+  #use list is a reference to pass
+  #check if node is empty, if yes, return the list
+  def postorder_helper(node, list)
+    if node.nil?
+      return list
+    end
+
+    list << {key: node.key, value: node.value}
+    postorder_helper(node.left, list)
+    postorder_helper(node.right, list)
+    
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: 0(n)
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  #--------------------------------------------------------------------------#
+  def height_helper(node)
+    if node.nil?
+      return 0
+    end
+    left_length = height_helper(node.left)
+    right_length = height_helper(node.right)
+
+    if left_length > right_length
+      longest_length = left_length
+    else
+      longest_length = right_length
+    end
+    longest_length += 1 # Add one for the current node
+    return longest_length
+
+  end
+  
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
👍 , the time & space complexities are O(n) for unbalanced trees and O(log n) when the trees are balanced.